File: android_gpt5/data/crypto/src/main/java/com/zarabandajose/transcripto/data/crypto/CaesarCipher.kt
package com.zarabandajose.transcripto.data.crypto

import com.zarabandajose.transcripto.domain.cipher.CipherMethod
import com.zarabandajose.transcripto.domain.model.Method

class CaesarCipher : CipherMethod {
    override val method: Method = Method.CAESAR

    private fun normalizeShift(shift: Int): Int {
        var s = shift % 26
        if (s < 0) s += 26
        return s
    }

    private fun hashSaltToShift(salt: ByteArray): Int {
        var h = 0
        for (b in salt) h = (31 * h + (b.toInt() and 0xFF))
        val mod = ((h % 26) + 26) % 26
        return mod
    }

    private fun transform(text: String, shift: Int): String {
        val s = normalizeShift(shift)
        val out = StringBuilder(text.length)
        for (c in text) {
            when (c) {
                in 'A'..'Z' -> {
                    val base = 'A'.code
                    val idx = c.code - base
                    val n = (idx + s) % 26
                    out.append((n + base).toChar())
                }
                in 'a'..'z' -> {
                    val base = 'a'.code
                    val idx = c.code - base
                    val n = (idx + s) % 26
                    out.append((n + base).toChar())
                }
                else -> out.append(c)
            }
        }
        return out.toString()
    }

    override fun encrypt(input: String, params: CipherMethod.Params): String {
    val baseShift = params.shift ?: error("Shift requerido")
    val salt = params.salt
    val extra = if (params.useSalt && salt != null) hashSaltToShift(salt) else 0
        return transform(input, baseShift + extra)
    }

    override fun decrypt(input: String, params: CipherMethod.Params): String {
    val baseShift = params.shift ?: error("Shift requerido")
    val salt = params.salt
    val extra = if (params.useSalt && salt != null) hashSaltToShift(salt) else 0
        return transform(input, -(baseShift + extra))
    }
}

File: android_flutter_claude/transcripto/packages/data_crypto/lib/src/ciphers/caesar_cipher.dart
import 'package:core_common/core_common.dart';
import 'package:domain/domain.dart';

/// Caesar cipher implementation
class CaesarCipher {
  const CaesarCipher();
  
  /// Encrypts text using Caesar cipher
  String encrypt(String input, CipherParams params) {
    final shift = _getEffectiveShift(params);
    return _processText(input, shift);
  }
  
  /// Decrypts text using Caesar cipher
  String decrypt(String input, CipherParams params) {
    final shift = _getEffectiveShift(params);
    return _processText(input, -shift);
  }
  
  int _getEffectiveShift(CipherParams params) {
    int baseShift = params.shift ?? 0;
    
    if (params.useSalt && params.salt != null) {
      final saltHash = Utils.simpleHash(params.salt!);
      final saltShift = saltHash % 26;
      baseShift = params.mode == CipherMode.encrypt 
          ? (baseShift + saltShift) % 26
          : (baseShift - saltShift) % 26;
    }
    
    return baseShift;
  }
  
  String _processText(String input, int shift) {
    final buffer = StringBuffer();
    
    for (int i = 0; i < input.length; i++) {
      final char = input[i];
      final code = char.codeUnitAt(0);
      
      if (code >= 65 && code <= 90) { // A-Z
        final shifted = ((code - 65 + shift) % 26 + 26) % 26;
        buffer.write(String.fromCharCode(shifted + 65));
      } else if (code >= 97 && code <= 122) { // a-z
        final shifted = ((code - 97 + shift) % 26 + 26) % 26;
        buffer.write(String.fromCharCode(shifted + 97));
      } else {
        buffer.write(char); // Keep non-alphabetic characters unchanged
      }
    }
    
    return buffer.toString();
  }
}
